workers/unity/Packages/com.improbable.gdk.core/GameObjectRepresentation/SpatialOSBehaviourManager.cs

```diff
@@ -1,60 +1,202 @@
 using System.Collections.Generic;
+using Improbable.Gdk.Core.MonoBehaviours;
 using Improbable.Worker;
 using UnityEngine;
+using Entity = Unity.Entities.Entity;
 
 namespace Improbable.Gdk.Core
 {
-    public class ReaderWriterImpl
-    {
-        // This is placeholder code.
-    }
-
     /// <summary>
-    ///     Invokes dispatcher callbacks on GameObjects that represent entities.
+    ///     Keeps track of Reader/Writer availability for SpatialOSBehaviours on a particular GameObject and decides when
+    ///     a SpatialOSBehaviour should be enabled, calling into the SpatialOSBehaviourLibrary for injection.
     /// </summary>
-    public class SpatialOSBehaviourManager
+    internal class SpatialOSBehaviourManager
     {
-        public GameObject GameObject;
-        private Dictionary<uint, List<MonoBehaviour>> componentIdToReaderSpatialOSBehaviours;
-        private Dictionary<uint, List<MonoBehaviour>> componentIdToWriterSpatialOSBehaviours;
-        private Dictionary<MonoBehaviour, int> numUnsatisfiedComponents;
-        private List<MonoBehaviour> spatialOSBehavioursToEnable;
-        private List<MonoBehaviour> spatialOSBehavioursToDisable;
-        private Dictionary<uint, Dictionary<MonoBehaviour, ReaderWriterImpl>> readersWriters;
+        private readonly Entity entity;
+        private readonly long spatialId;
+
+        private readonly Dictionary<uint, HashSet<MonoBehaviour>> componentIdToReaderSpatialOSBehaviours
+            = new Dictionary<uint, HashSet<MonoBehaviour>>();
+
+        private readonly Dictionary<uint, HashSet<MonoBehaviour>> componentIdToWriterSpatialOSBehaviours
+            = new Dictionary<uint, HashSet<MonoBehaviour>>();
+
+        private readonly Dictionary<MonoBehaviour, int> numUnsatisfiedReadersOrWriters
+            = new Dictionary<MonoBehaviour, int>();
+
+        private readonly Dictionary<MonoBehaviour, Dictionary<uint, IReaderInternal>> behaviourToCreatedReadersWriters
+            = new Dictionary<MonoBehaviour, Dictionary<uint, IReaderInternal>>();
+
+        private readonly HashSet<MonoBehaviour> spatialOSBehavioursToEnable = new HashSet<MonoBehaviour>();
+        private readonly HashSet<MonoBehaviour> spatialOSBehavioursToDisable = new HashSet<MonoBehaviour>();
 
-        public SpatialOSBehaviourManager(GameObject gameObject)
+        private readonly Dictionary<uint, HashSet<IReaderInternal>> componentIdToCreatedReadersWriters =
```

Comment:
There's a lot of bookkeeping and cross-referencing going on here. Consider refactoring some of this into structs to reduce to number of raw mappings that you have here.